File: src/Day2/Part2.hs
module Day2.Part2 where

import Day2.Util ( Entry (..), parseEntry )

day2part2 :: String -> String
day2part2 input = show $ solve entries
  where entries = map parseEntry $ lines input

solve :: [Entry] -> Int
solve = length . filter isValid

isValid :: Entry -> Bool
isValid entry = (password !! pos1 == char) /= (password !! pos2 == char)
  where pos1 = a entry - 1
        pos2 = b entry - 1
        char = c entry
        password = p entry

File: src/Day2/part2.rs
use crate::day2::util::{Entry, parse_entries};

fn is_valid(entry: &Entry) -> bool {
    let bytes = entry.p.as_bytes();
    let a = bytes[entry.a - 1] as char;
    let b = bytes[entry.b - 1] as char;
    return (a == entry.c) != (b == entry.c);
}

pub fn day2part2(input: &String) -> Option<String> {
    let entries = match parse_entries(input) {
        Some(e) => e,
        None => return None,
    };
    let valid_entries: Vec<&Entry> = entries
        .iter()
        .filter(|e| is_valid(e))
        .collect();
    return Some(valid_entries.len().to_string());
}
